// voyageai/dist/cjs/serialization/client/requests/EmbedRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbedRequest = void 0;
const core = __importStar(require("../../../core/index.js"));
const EmbedRequestEncodingFormat_js_1 = require("../../types/EmbedRequestEncodingFormat.js");
const EmbedRequestInput_js_1 = require("../../types/EmbedRequestInput.js");
const EmbedRequestInputType_js_1 = require("../../types/EmbedRequestInputType.js");
const EmbedRequestOutputDtype_js_1 = require("../../types/EmbedRequestOutputDtype.js");
exports.EmbedRequest = core.serialization.object({
    input: EmbedRequestInput_js_1.EmbedRequestInput,
    model: core.serialization.string(),
    inputType: core.serialization.property("input_type", EmbedRequestInputType_js_1.EmbedRequestInputType.optional()),
    truncation: core.serialization.boolean().optional(),
    encodingFormat: core.serialization.property("encoding_format", EmbedRequestEncodingFormat_js_1.EmbedRequestEncodingFormat.optional()),
    outputDimension: core.serialization.property("output_dimension", core.serialization.number().optional()),
    outputDtype: core.serialization.property("output_dtype", EmbedRequestOutputDtype_js_1.EmbedRequestOutputDtype.optional()),
});
