// voyageai/dist/cjs/local/model-registry.js
"use strict";
/**
 * Model configuration and registry for local models.
 * Mirrors Python SDK's voyageai/local/model_registry.py
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUPPORTED_MODELS = void 0;
exports.isLocalModel = isLocalModel;
exports.getModelConfig = getModelConfig;
exports.validateDimension = validateDimension;
exports.validatePrecision = validatePrecision;
exports.SUPPORTED_MODELS = {
    "voyage-4-nano": {
        onnxModelId: "onnx-community/voyage-4-nano-ONNX",
        defaultDimension: 2048,
        supportedDimensions: [2048, 1024, 512, 256],
        supportedPrecisions: ["float32", "int8", "uint8", "binary", "ubinary"],
        maxTokens: 32768,
    },
};
function isLocalModel(model) {
    return model in exports.SUPPORTED_MODELS;
}
function getModelConfig(model) {
    const config = exports.SUPPORTED_MODELS[model];
    if (!config) {
        throw new Error(`Unsupported local model '${model}'. ` +
            `Supported models: ${Object.keys(exports.SUPPORTED_MODELS).join(", ")}`);
    }
    return config;
}
function validateDimension(config, dimension) {
    if (dimension == null) {
        return config.defaultDimension;
    }
    if (!config.supportedDimensions.includes(dimension)) {
        throw new Error(`Invalid outputDimension ${dimension}. ` +
            `Supported dimensions: ${config.supportedDimensions.join(", ")}`);
    }
    return dimension;
}
function validatePrecision(config, precision) {
    if (precision == null) {
        return undefined;
    }
    if (!config.supportedPrecisions.includes(precision)) {
        throw new Error(`Invalid outputDtype '${precision}'. ` +
            `Supported dtypes: ${config.supportedPrecisions.join(", ")}`);
    }
    return precision;
}
