// voyageai/dist/cjs/extended/ExtendedClient.js
"use strict";
/**
 * Extended VoyageAI client with local model support and client-side chunking
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoyageAIClient = void 0;
exports.buildContextualizedResult = buildContextualizedResult;
exports.validateAndNormalizeContextualizedInputs = validateAndNormalizeContextualizedInputs;
exports.applyChunking = applyChunking;
exports.defaultChunkFn = defaultChunkFn;
const Client_js_1 = require("../Client.js");
const HttpResponsePromise_js_1 = require("../core/fetcher/HttpResponsePromise.js");
const RawResponse_js_1 = require("../core/fetcher/RawResponse.js");
const index_js_1 = require("../local/index.js");
const tokenizer_js_1 = require("../local/tokenizer.js");
class VoyageAIClient extends Client_js_1.VoyageAIClient {
    /**
     * Voyage embedding endpoint receives as input a string (or a list of strings) and other arguments such as the preferred model name, and returns a response containing a list of embeddings.
     *
     * For local models (e.g., 'voyage-4-nano'), embeddings are computed locally using Transformers.js.
     * For all other models, the request is forwarded to the Voyage AI API.
     *
     * @param {VoyageAI.EmbedRequest} request
     * @param {GeneratedClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     // Local embedding (no API key required)
     *     await client.embed({
     *         input: "hello world",
     *         model: "voyage-4-nano"
     *     })
     *
     * @example
     *     // API embedding
     *     await client.embed({
     *         input: "hello world",
     *         model: "voyage-3-large"
     *     })
     */
    embed(request, requestOptions) {
        if ((0, index_js_1.isLocalModel)(request.model)) {
            return HttpResponsePromise_js_1.HttpResponsePromise.fromPromise((0, index_js_1.localEmbed)(request).then((data) => ({ data, rawResponse: RawResponse_js_1.unknownRawResponse })));
        }
        return super.embed(request, requestOptions);
    }
    /**
     * Contextualized embeddings with client-side chunking and server-side auto-chunking support.
     *
     * Returns a `ContextualizedEmbedResult` wrapping the API response with extracted
     * embeddings, chunk texts, and token counts — mirroring the Python SDK's
     * `ContextualizedEmbeddingsObject`.
     *
     * @param request - Embed request with optional `chunkFn` for client-side chunking.
     * @param requestOptions - Request-specific configuration.
     */
    // @ts-expect-error Return type narrowed from HttpResponsePromise to wrapped result
    contextualizedEmbed(request, requestOptions) {
        const _super = Object.create(null, {
            contextualizedEmbed: { get: () => super.contextualizedEmbed }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const { chunkFn } = request, apiRequest = __rest(request, ["chunkFn"]);
            const normalizedRequest = validateAndNormalizeContextualizedInputs(apiRequest, chunkFn);
            const requestInputs = chunkFn
                ? applyChunking(normalizedRequest.inputs, chunkFn)
                : normalizedRequest.inputs;
            const response = yield _super.contextualizedEmbed.call(this, Object.assign(Object.assign({}, normalizedRequest), { inputs: requestInputs }), requestOptions);
            const clientChunkTexts = chunkFn ? requestInputs : undefined;
            return buildContextualizedResult(response, clientChunkTexts);
        });
    }
    /**
     * Tokenize texts using the specified model's tokenizer.
     * Works with both local models (e.g., 'voyage-4-nano') and API models (e.g., 'voyage-3-large').
     *
     * @param texts - Array of strings to tokenize
     * @param model - Voyage model name
     * @returns Array of TokenizeResult, each containing tokens (string[]) and ids (number[])
     *
     * @example
     *     const results = await client.tokenize(["hello world"], "voyage-4-nano");
     *     console.log(results[0].tokens); // ["hello", "world"]
     *     console.log(results[0].ids);    // [123, 456]
     */
    tokenize(texts, model) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, tokenizer_js_1.tokenizeTexts)(model, texts);
        });
    }
}
exports.VoyageAIClient = VoyageAIClient;
function buildContextualizedResult(response, clientChunkTexts) {
    var _a, _b, _c, _d, _e;
    const results = [];
    const serverTextsPerDoc = [];
    for (let i = 0; i < ((_a = response.data) !== null && _a !== void 0 ? _a : []).length; i++) {
        const doc = response.data[i];
        const embeddings = ((_b = doc.data) !== null && _b !== void 0 ? _b : []).map((chunk) => { var _a; return (_a = chunk.embedding) !== null && _a !== void 0 ? _a : []; });
        let chunkTexts;
        if (clientChunkTexts !== undefined) {
            chunkTexts = clientChunkTexts[i];
        }
        else {
            const perDocTexts = ((_c = doc.data) !== null && _c !== void 0 ? _c : []).map((chunk) => chunk.text);
            if (perDocTexts.every((t) => t !== undefined)) {
                serverTextsPerDoc.push(perDocTexts);
                chunkTexts = perDocTexts;
            }
            else if (perDocTexts.some((t) => t !== undefined)) {
                throw new Error(`inputs[${i}] returned a partial set of chunk texts; expected text on every chunk or none`);
            }
            else {
                serverTextsPerDoc.push(undefined);
                chunkTexts = undefined;
            }
        }
        results.push({ index: i, embeddings, chunkTexts });
    }
    let resultChunkTexts;
    if (clientChunkTexts !== undefined) {
        resultChunkTexts = clientChunkTexts;
    }
    else if (serverTextsPerDoc.length > 0) {
        const populated = serverTextsPerDoc.filter((t) => t !== undefined);
        if (populated.length > 0 && populated.length !== serverTextsPerDoc.length) {
            throw new Error("response returned chunk texts for some documents but not others; expected all-or-nothing");
        }
        if (populated.length > 0) {
            resultChunkTexts = populated;
        }
    }
    return {
        results,
        totalTokens: (_e = (_d = response.usage) === null || _d === void 0 ? void 0 : _d.totalTokens) !== null && _e !== void 0 ? _e : 0,
        chunkTexts: resultChunkTexts,
        chunkerVersion: response.chunkerVersion,
        rawResponse: response,
    };
}
function isFlatStringArray(inputs) {
    return inputs.length > 0 && inputs.every((item) => typeof item === "string");
}
function validateAndNormalizeContextualizedInputs(request, chunkFn) {
    const { inputs, inputType, enableAutoChunking, chunkSize, chunkOverlap } = request;
    if (chunkFn !== undefined && enableAutoChunking) {
        throw new Error("chunkFn cannot be combined with enableAutoChunking: true");
    }
    const hasChunkSize = chunkSize !== undefined && chunkSize !== null;
    const hasChunkOverlap = chunkOverlap !== undefined && chunkOverlap !== null;
    if (!enableAutoChunking && (hasChunkSize || hasChunkOverlap)) {
        throw new Error("chunkSize and chunkOverlap require enableAutoChunking: true");
    }
    if (hasChunkSize && hasChunkOverlap && chunkOverlap >= chunkSize) {
        throw new Error(`chunkOverlap (${chunkOverlap}) must be less than chunkSize (${chunkSize})`);
    }
    if (hasChunkSize && chunkSize < 1) {
        throw new Error("chunkSize must be greater than or equal to 1");
    }
    if (hasChunkOverlap && chunkOverlap < 0) {
        throw new Error("chunkOverlap must be greater than or equal to 0");
    }
    if (hasChunkOverlap && !hasChunkSize) {
        throw new Error("chunkOverlap requires chunkSize");
    }
    if (!inputs || inputs.length === 0) {
        throw new Error("inputs must not be empty");
    }
    let normalizedInputs;
    if (isFlatStringArray(inputs)) {
        if (inputType !== "query" && !enableAutoChunking) {
            throw new Error("Flat string[] inputs require enableAutoChunking: true or inputType: 'query'");
        }
        normalizedInputs = inputs.map((s) => [s]);
    }
    else {
        normalizedInputs = inputs;
    }
    if (enableAutoChunking) {
        if (inputType !== "document") {
            throw new Error("enableAutoChunking: true requires inputType: 'document'");
        }
        for (let i = 0; i < normalizedInputs.length; i++) {
            if (normalizedInputs[i].length !== 1) {
                throw new Error(`inputs[${i}] has ${normalizedInputs[i].length} chunks; auto-chunking expects one string per document`);
            }
        }
    }
    return Object.assign(Object.assign({}, request), { inputs: normalizedInputs });
}
function applyChunking(inputs, chunkFn) {
    return inputs.map((doc) => doc.flatMap((text) => chunkFn(text)));
}
const DEFAULT_CHUNK_SIZE = 2048;
const SEPARATORS = [
    "\n\n",
    "\n",
    "．", // Fullwidth full stop
    "。", // Ideographic full stop
    "，", // Fullwidth comma
    "、", // Ideographic comma
    ".",
    ",",
    " ",
    "​", // Zero-width space
    "",
];
function recursiveSplit(text, separators, chunkSize) {
    if (text.length <= chunkSize) {
        return [text];
    }
    const sep = separators[0];
    const remainingSeparators = separators.slice(1);
    if (sep === undefined || sep === "") {
        const chunks = [];
        for (let i = 0; i < text.length; i += chunkSize) {
            chunks.push(text.slice(i, i + chunkSize));
        }
        return chunks;
    }
    const parts = text.split(sep);
    const chunks = [];
    let current = "";
    for (let i = 0; i < parts.length; i++) {
        const part = parts[i];
        const piece = i < parts.length - 1 ? part + sep : part;
        if (current.length + piece.length <= chunkSize) {
            current += piece;
        }
        else {
            if (current.length > 0) {
                chunks.push(current);
            }
            if (piece.length <= chunkSize) {
                current = piece;
            }
            else {
                const subChunks = recursiveSplit(piece, remainingSeparators, chunkSize);
                chunks.push(...subChunks.slice(0, -1));
                current = subChunks[subChunks.length - 1];
            }
        }
    }
    if (current.length > 0) {
        chunks.push(current);
    }
    return chunks;
}
function addOverlap(chunks, overlap) {
    if (overlap === 0 || chunks.length <= 1)
        return chunks;
    const result = [chunks[0]];
    for (let i = 1; i < chunks.length; i++) {
        const prev = chunks[i - 1];
        const overlapText = prev.slice(-overlap);
        result.push(overlapText + chunks[i]);
    }
    return result;
}
function defaultChunkFn(chunkSize = DEFAULT_CHUNK_SIZE, chunkOverlap = 0) {
    return (text) => {
        if (text.length === 0) {
            return [text];
        }
        const chunks = recursiveSplit(text, SEPARATORS, chunkSize);
        return addOverlap(chunks, chunkOverlap);
    };
}
