// voyageai/dist/cjs/extended/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAndNormalizeContextualizedInputs = exports.VoyageAIClient = exports.defaultChunkFn = exports.buildContextualizedResult = exports.applyChunking = exports.localEmbed = exports.isLocalModel = exports.GeneratedVoyageAIClient = void 0;
// Re-export everything from generated SDK
__exportStar(require("../api/index.js"), exports);
// For advanced users who want the original generated client
var Client_js_1 = require("../Client.js");
Object.defineProperty(exports, "GeneratedVoyageAIClient", { enumerable: true, get: function () { return Client_js_1.VoyageAIClient; } });
__exportStar(require("../errors/index.js"), exports);
// Export local utilities for advanced use cases
var index_js_1 = require("../local/index.js");
Object.defineProperty(exports, "isLocalModel", { enumerable: true, get: function () { return index_js_1.isLocalModel; } });
Object.defineProperty(exports, "localEmbed", { enumerable: true, get: function () { return index_js_1.localEmbed; } });
// Export our extended client as the main one
var ExtendedClient_js_1 = require("./ExtendedClient.js");
Object.defineProperty(exports, "applyChunking", { enumerable: true, get: function () { return ExtendedClient_js_1.applyChunking; } });
Object.defineProperty(exports, "buildContextualizedResult", { enumerable: true, get: function () { return ExtendedClient_js_1.buildContextualizedResult; } });
Object.defineProperty(exports, "defaultChunkFn", { enumerable: true, get: function () { return ExtendedClient_js_1.defaultChunkFn; } });
Object.defineProperty(exports, "VoyageAIClient", { enumerable: true, get: function () { return ExtendedClient_js_1.VoyageAIClient; } });
Object.defineProperty(exports, "validateAndNormalizeContextualizedInputs", { enumerable: true, get: function () { return ExtendedClient_js_1.validateAndNormalizeContextualizedInputs; } });
