// voyageai/dist/esm/local/local-embedder.mjs
/**
 * Local embedding using @huggingface/transformers (ONNX Runtime).
 * Runs voyage-4-nano entirely locally - no API calls after initial model download.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { getModelConfig, validateDimension, validatePrecision } from "./model-registry.mjs";
import { getTransformers, getTokenizerForModel } from "./tokenizer.mjs";
// Cached pipelines (persist across calls)
const pipelineCache = new Map();
function getPipeline(onnxModelId) {
    return __awaiter(this, void 0, void 0, function* () {
        let extractor = pipelineCache.get(onnxModelId);
        if (!extractor) {
            const { pipeline } = yield getTransformers();
            extractor = yield pipeline("feature-extraction", onnxModelId, { dtype: "fp32" });
            pipelineCache.set(onnxModelId, extractor);
        }
        return extractor;
    });
}
function normalize(vec) {
    const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0));
    if (norm === 0)
        return vec;
    return vec.map((v) => v / norm);
}
export function localEmbed(request) {
    return __awaiter(this, void 0, void 0, function* () {
        const config = getModelConfig(request.model);
        const dimension = validateDimension(config, request.outputDimension);
        validatePrecision(config, request.outputDtype);
        const extractor = yield getPipeline(config.onnxModelId);
        const inputs = Array.isArray(request.input) ? request.input : [request.input];
        const embeddings = [];
        for (const text of inputs) {
            const output = yield extractor(text, { pooling: "mean", normalize: true });
            // output is a Tensor; convert to flat array then truncate to requested dimension
            const fullVec = Array.from(output.data).slice(0, output.dims[1]);
            const truncated = fullVec.slice(0, dimension);
            // Re-normalize after Matryoshka truncation
            embeddings.push(normalize(truncated));
        }
        // Token counting via tokenizer
        const tokenizer = yield getTokenizerForModel(request.model);
        let totalTokens = 0;
        for (const text of inputs) {
            const encoded = tokenizer(text);
            totalTokens += encoded.input_ids.size;
        }
        return {
            object: "list",
            model: request.model,
            data: embeddings.map((embedding, idx) => ({
                object: "embedding",
                embedding,
                index: idx,
            })),
            usage: { totalTokens },
        };
    });
}
