// voyageai/dist/esm/serialization/client/requests/ContextualizedEmbedRequest.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../core/index.mjs";
import { ContextualizedEmbedRequestInputs } from "../../types/ContextualizedEmbedRequestInputs.mjs";
import { ContextualizedEmbedRequestInputType } from "../../types/ContextualizedEmbedRequestInputType.mjs";
import { ContextualizedEmbedRequestOutputDtype } from "../../types/ContextualizedEmbedRequestOutputDtype.mjs";
export const ContextualizedEmbedRequest = core.serialization.object({
    inputs: ContextualizedEmbedRequestInputs,
    model: core.serialization.string(),
    inputType: core.serialization.property("input_type", ContextualizedEmbedRequestInputType.optional()),
    outputDimension: core.serialization.property("output_dimension", core.serialization.number().optional()),
    outputDtype: core.serialization.property("output_dtype", ContextualizedEmbedRequestOutputDtype.optional()),
    enableAutoChunking: core.serialization.property("enable_auto_chunking", core.serialization.boolean().optional()),
    chunkSize: core.serialization.property("chunk_size", core.serialization.number().optional()),
    chunkOverlap: core.serialization.property("chunk_overlap", core.serialization.number().optional()),
});
