// voyageai/dist/esm/serialization/client/requests/EmbedRequest.mjs
// This file was auto-generated by Fern from our API Definition.
import * as core from "../../../core/index.mjs";
import { EmbedRequestEncodingFormat } from "../../types/EmbedRequestEncodingFormat.mjs";
import { EmbedRequestInput } from "../../types/EmbedRequestInput.mjs";
import { EmbedRequestInputType } from "../../types/EmbedRequestInputType.mjs";
import { EmbedRequestOutputDtype } from "../../types/EmbedRequestOutputDtype.mjs";
export const EmbedRequest = core.serialization.object({
    input: EmbedRequestInput,
    model: core.serialization.string(),
    inputType: core.serialization.property("input_type", EmbedRequestInputType.optional()),
    truncation: core.serialization.boolean().optional(),
    encodingFormat: core.serialization.property("encoding_format", EmbedRequestEncodingFormat.optional()),
    outputDimension: core.serialization.property("output_dimension", core.serialization.number().optional()),
    outputDtype: core.serialization.property("output_dtype", EmbedRequestOutputDtype.optional()),
});
